docs/design/jet/025-jet-job-member-pruning.md

```diff
@@ -0,0 +1,205 @@
+to shorten their meaning we mark them as 'heavyweight' and 'lightweight'. The difference between these two approaches 
+is an exact stage when we understand that some member may be pruned from job execution: 'heavyweight' approach tries 
+to be more generic and define it for any accept DAG, when 'lightweight' approach focus on hinting `ExecutionPlanBuilder`
+with additional meta-information from DAG and JobConfig constructed by SQL optimizer.
+
+##### Heavyweight approach
+
+The core idea here is that  `ExecutionPlanBuilder` would have detailed analysis of received DAG before execution plan
+creation and then, possibly, trying to optimize DAG and only then apply member pruning based. The requirements
+to prune member are :
+- each vertex in the DAG may work without an input;
+- DAG does not contain `distributed-broadcast` edges;
+
+As a pros, using this approach generify member pruning usage for both SQL, Pipeline API; allow usage for any connectors.
+As a cons, this approach brings high complexity in problem analysis, long development and big list of corner cases, which
+are discussed in chapter below.
+
+##### Lightweight approach
+
+Unlike a way described above, we can think that member pruning would be beneficial mostly for small jobs.
+The overwhelming majority of such jobs are generated by SQL, and our team effort is to replace Predicate API,
+where `PartitionPredicate` is available for partitioned data querying.
+For that, we can use SQL optimization phase also for determining partitions and for all relations: we extract information
+about required partitions to run to JobConfig, and then mark DAG as 'ableToPruneMembers'. This mark is an indicator
+for `ExecutionPlanBuilder` to choose code path with member pruning during execution plans creation.
+As a helpful API for SQL's `CreateDagVisitor`, we introduced new wrapper for both `ProcessorMetaSupplier` and
+`ProcessorSupplier` - `memberPruningProcessorMetaSupplier`, where `doesWorkWithoutInput` method returns `false`. This PMS
+make easier and more precise choosing members to prune.
+
+### Processor pruning
+
+For processor pruning we would like to separate two kinds of pruning: **intra-member** processor pruning, which
+is applicable only for processor creation elimination inside one member, and **inter-member** processor pruning.
+
+#### Use cases for processor pruning
+
+##### Processor pruning for Scan -> Transform -> Aggregate
+
+1. Single-staged aggregation.
+
+![Scan + Scan -> HashJoin](https://svgshare.com/i/tTM.svg)
+
+On the picture above we can see optimized example from member pruning case. We can go further and just don't create 
+processors which are not participating in data processing by tuning local parallelism parameter with 
+partition involvement knowledge : 
+
+![Scan + Scan -> HashJoin](https://svgshare.com/i/tUD.svg)
+
+So, it's a good target for **intra-member** processor pruning.
```

Comment:
**intra-member** processor pruning -> local parallelism adjustement

but there is also case of completely removing processor given processor instance but not removing their dependent processors because they `doWorkWithoutInput`